src/main/kotlin/app/RepoHasher.kt

```diff
@@ -0,0 +1,89 @@
+// Copyright 2017 Sourcerer Inc. All Rights Reserved.
+// Author: Anatoly Kislov (anatoly@sourcerer.io)
+
+package app
+
+import app.model.Commit
+import app.model.Repo
+import app.utils.RepoHelper
+import io.reactivex.Observable
+import org.eclipse.jgit.api.Git
+import org.eclipse.jgit.revwalk.RevCommit
+import org.eclipse.jgit.revwalk.RevWalk
+import java.io.File
+import java.io.IOException
+
+/**
+ * RepoHasher hashes repository and uploads stats to server.
+ */
+class RepoHasher(val repo: Repo) {
+    fun update() {
+        //TODO(anatoly): Implement repository analysis.
+        //TODO(anatoly): Implement data transfer.
+
+        var firstCommit: RevCommit
+
+        // Get commits loading observable and wait for subscribers.
+        val observableRevCommits = getObservableRevCommits().publish()
+
+        // Main data flow. Read commits from repo and send them to server.
+        observableRevCommits
+                .doOnNext {
+                    Logger.info("Commit: ${it.id.name} ${it.shortMessage}")
+                }
+                .map { c -> Commit(c) }
+                .subscribe(
+                        {
+                            Logger.debug("onNext")
+                            SourcererApi.postCommitAsync(Commit(), {}, {})
+                        },
+                        { e ->
+                            Logger.error("Error while hashing: ${e.message}")
+                        },
+                        {
+                            Logger.debug("onComplete")
+                        }
+                )
+
+        // Additional data flow. Read first commit.
+        observableRevCommits
+                .take(1)
+                .subscribe({
+                    firstCommit = it
+                }, {
+                    // OnError processed on main data flow. Need to be defined.
+                })
+
+        // Start emitting commits.
+        observableRevCommits.connect()
+    }
+
+    fun getObservableRevCommits(): Observable<RevCommit> = Observable.create {
+        subscriber ->
+
+        val git = try {
+            Git.open(File(repo.path))
+        } catch (e: IOException) {
+            subscriber.onError(e)
+            null
+        }
+
+        if (git != null) {
+            val repository = git.repository
+            try {
+                val revWalk = RevWalk(repository)
+                val commitId = repository.resolve("refs/heads/master")
```

Comment:
This is a very much a shotgun approach. We need to be careful here, and only take commits that belong to a particular user. We should look into repo/.git/config and ~/.gitconfig for user's emails in those repos, and use those to pick commits. Also, later on, there will be a need to generate profiles for user's colleagues (i.e. people the user actively works with), so this needs to be organized in a way that one iteration over commits targets a specific user identified by a set of email addresses.

At the very least, add a TODO.